zk/groth16/src/verifier.rs

```diff
@@ -14,6 +21,58 @@ pub fn groth16_verify<E: Pairing>(
     Groth16::<E, LibsnarkReduction>::verify_proof(vk.as_ref(), &proof, public_inputs)
 }
 
+pub fn groth16_batch_verify(
+    vk: &PreparedVerificationKey<Bn254>,
+    proofs: &[Proof<Bn254>],
+    public_inputs: &[Vec<Fr>],
+) -> R1CSResult<bool> {
+    let mut rng = thread_rng();
+    let r = Fr::rand(&mut rng);
+    let r_roots = compute_r_powers(r, proofs.len());
+    let r_sum: Fr = r_roots.iter().sum();
+
+    let pis_c: Vec<G1Affine> = proofs.iter().map(|proof| proof.pi_c).collect();
+
+    let batched_pi_c = G1Projective::msm(&pis_c, &r_roots).unwrap().into_affine();
+
+    let batched_public_inputs: Vec<Fr> = std::iter::once(r_sum)
+        .chain((0..public_inputs[0].len()).map(|i| {
```

Comment:
This use of `public_inputs[0].len()` here seems unwise, idk if it can be exploited, but generally we shouldn't be conditioning verification logic on untrusted inputs if we can derive these values from trusted sources. Can we get this length from the verification key (something like `vk.ic.len() - 1`)?

```rust
        .chain((0..public_inputs[0].len()).map(|i| {
```